internal/controllers/upboundagent/reconciler.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2021 Upbound Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package upboundagent
+
+import (
+	"context"
+	"time"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	internalmeta "github.com/upbound/universal-crossplane/internal/meta"
+)
+
+const (
+	reconcileTimeout = 1 * time.Minute
+
+	deploymentUpboundAgent = "upbound-agent"
+	keyToken               = "token"
+)
+
+const (
+	errGetSecret              = "failed to get control plane token secret"
+	errNoTokenInSecret        = "secret %s does not contain a token for key \"%s\""
+	errFailedToSyncDeployment = "failed to sync agent deployment"
+)
+
+var (
+	secretsKind     = schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Secret"}
+	deploymentsKind = schema.GroupVersionKind{Group: "apps", Version: "v1", Kind: "Deployment"}
```

Comment:
nitpick: might make sense to access these by importing the packages, but in practice these won't be changing (without major implications), so feel free to ignore suggestion if you prefer as-is :+1:

Reply:
Frankly, I couldn't find them in k8s packages. But, as you pointed current implementation should be ok given there is no (real) possibility for them to get changed.